### PR TITLE
Implement Bezier Element export for ParaView

### PR DESCRIPTION
**Pull request by ckarampa**
VTK (and subsequently ParaView) now natively supports Bezier elements. This allows us to export geometries ( and solution fields in a future PR) by means of Bezier extraction.

### NEW:
- `gsWriteParaviewBezier` accepts a `gsMultiPatch` and exports it as a *single* `.vtu` file containing `VTK_BEZIER_QUADRILATERAL` elements. Currently this is implemented only for bivariate geometries.
- `gsMultiPatch::BezierExtraction()` implements Bezier extraction via quasi interpolation and has been brought over from `gismo/gsLsDyna` and returns a multi patch consisting of all the bezier elements.. The struct `ElementBlock` and method `BezierOperator` are necessary for its  operation. The struct has been 'hidden' inside the `internal` namespace and `BezierOperator` has been marked private. 

### IMPROVED:
- Static functions of `gsParaviewDataSet` moved to a separate file, `gsParaviewUtils`. These deal mainly with formatting data to the .vtk format.
- `toDataArray` now also supports integers, matrices of integers, and custom xml tag attribute-value pairs.
- A flag has been added to `inputOutput_example`, which exports the geometry as bezier elements.
- `filedata/surfaces/teapot.xml` now contains a multi patch definition.

### TODO:
- [ ] To avoid compilation errors the PR in `gismo/gsLsDyna` which deletes the struct/funtions from there needs to be accepted along with this PR.

# Please consider the following checklist before issuing a pull
request:
- [x] Have you added an explanation of what your changes do and why
  you'd like us to include them?
- [x] Have you documented any new codes using Doxygen comments?
- [x] Have you written new tests or examples for your changes?
-----

### CODE REVIEW

File: src/gsCore/gsMultiPatch.h [hunk 1]
```diff
@@ -416,6 +430,8 @@ class gsMultiPatch : public gsBoxTopology, public gsFunctionSet<T>
     const InterfaceRep & interfaceRep() const { return m_ifaces; }
     const BoundaryRep & boundaryRep() const { return m_bdr; }
     const BoundaryRep & sides() const { return m_sides; }
+
+    gsMultiPatch<T> BezierExtraction() const;
```
Comment (filiatra) [inline src/gsCore/gsMultiPatch.h]: rename as:
gsMultiPatch<T> extractBezier() const;

add one line of doxygen doc

File: src/gsCore/gsMultiPatch.hpp [hunk 2]
```diff
@@ -994,5 +998,107 @@ void gsMultiPatch<T>::constructSides()
...
+
+        gsKnotVector<> kv1(0,1,0,ElBlock.PR+1); 
+        gsKnotVector<> kv2(0,1,0,ElBlock.PS+1);
+        // coefs.setZero( (ElBlock.PR+1)*(ElBlock.PS+1), controlPoints.cols()-1 );
+
+        // Loop over all elements of the Element Block
+        auto Ait = ElBlock.actives.begin();        // Actives Iterator
+        auto Cit = ElBlock.coefVectors.begin();    // Coefficients Iteratos
+
+        gsMatrix<real_t> newCoefs( ((ElBlock.PR+1) * (ElBlock.PS+1)), targetDim() );
```
Comment (filiatra) [inline src/gsCore/gsMultiPatch.hpp]: newCoefs is not used ?

File: src/gsCore/gsMultiPatch.h [hunk 3]
```diff
@@ -448,6 +464,8 @@ class gsMultiPatch : public gsBoxTopology, public gsFunctionSet<T>
            const gsVector<bool> &matched,
            gsVector<index_t> &dirMap, gsVector<bool>    &dirO,
            T tol, index_t reference=0);
+    
+    std::map<index_t, internal::ElementBlock> BezierOperator() const;
```
Comment (filiatra) [inline src/gsCore/gsMultiPatch.h]: add one line of doxygen doc

File: src/gsCore/gsMultiPatch.hpp [hunk 4]
```diff
@@ -994,5 +998,107 @@ void gsMultiPatch<T>::constructSides()
...
+    index_t NN; // Number of control points of the Bezier element
+    gsMatrix<index_t> actives; // Active basis functions
+    index_t offset=0; // Offset of active's numbering ( local id -> multipatch id)
+
+    for (index_t p=0; p<nPatches(); ++p)
+    {
+        gsBasis<T> * basis = & patch(p).basis();
+
+        // Create the Bezier Basis
+        gsKnotVector<T> kv1;
```
Comment (filiatra) [inline src/gsCore/gsMultiPatch.hpp:1026]: Make the basis creation using static dispatch here

File: src/gsCore/gsMultiPatch.hpp [hunk 5]
```diff
@@ -994,5 +998,107 @@ void gsMultiPatch<T>::constructSides()
...
+{
+    GISMO_ENSURE( 2==domainDim(), "Anything other than bivariate splines is not yet supported!");
+    std::map<index_t, internal::ElementBlock> ElementBlocks = BezierOperator();
+    gsMultiPatch<T> result;
+
+    for (auto const& pair : ElementBlocks)
+    {
+        internal::ElementBlock ElBlock = pair.second;
+
+        gsKnotVector<> kv1(0,1,0,ElBlock.PR+1);
```
Comment (filiatra) [inline src/gsCore/gsMultiPatch.hpp:1114]: Make the basis creation using static dispatch here as well

File: src/gsIO/gsParaviewUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+#define VTK_BEZIER_QUADRILATERAL 77
+
+
+namespace gismo
+{
+    gsMatrix<real_t> vtkIDTransform(index_t nU, index_t nV)
+    {
+        // T converts coefs from G+Smo's convetnion to ParaView's convention
+        gsMatrix<real_t> T(nU*nV, nU*nV);
```
Comment (filiatra) [inline src/gsIO/gsParaviewUtils.cpp]: Rename the matrix T not to confuse with the template parameter

File: src/gsIO/gsParaviewUtils.h [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+
+
+
+
+    /// @brief ID transformation between G+Smo and vtk  control point IDs 
+    /// @param nU Number of control points in u parametric direction
+    /// @param nV Number of control points in u parametric direction
+    /// @return 
+    gsMatrix<real_t> vtkIDTransform(index_t nU, index_t nV);
```
Comment (filiatra) [inline src/gsIO/gsParaviewUtils.h]: needs GISMO_EXPORT attribute (Because it is not templated)

File: src/gsIO/gsParaviewUtils.h [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+// #include <gismo.h>
+#include <gsCore/gsForwardDeclarations.h>
+#include <gsMSplines/gsMappedBasis.h>   // Only to make linker happy
+#include <gsCore/gsDofMapper.h>         // Only to make linker happy
+#include <gsAssembler/gsExprHelper.h>  
+#include <gsAssembler/gsExprEvaluator.h>
+// #include <gsIO/gsIOUtils.h>
+#include <gsCore/gsLinearAlgebra.h>
+#include <gsIO/gsBase64.h>
+
```
Comment (filiatra) [inline src/gsIO/gsParaviewUtils.h:23]: too many includes, they should move to the hpp file

File: src/gsIO/gsParaviewUtils.h [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+
+    /// @brief Converts an integer to a 'DataArray' xml tag, which is returned as a string.
+    /// @param num The integer to be formatted
+    /// @param attributes Optional, map of strings, with attribute name mapping to attribute value.
+    /// @param ind Optional, indentation level for the resulting string.
+    /// @return 
+    std::string toDataArray(index_t num, std::map<std::string, std::string> attributes={{"",""}});
+
```
Comment (filiatra) [inline src/gsIO/gsParaviewUtils.h:135]: needs GISMO_EXPORT

**Review by filiatra — CHANGES_REQUESTED** [review]

**Comment by ckarampa** [comment]
In the `gsMultipatch::BezierOperator` we are grouping the `ElementBlocks` by `NN` i.e. the number of control points of each Bezier element. This will be problematic if in a multipatch there exist patches of different degree that have the same number of control points (e.g. 2x2 & 4x1). This might be rare but it is not impossible. We could group them by `(PR, PS, PT)` ( the degree per direction) to avoid this.

**Comment by ckarampa** [comment]
> In the `gsMultipatch::BezierOperator` we are grouping the `ElementBlocks` by `NN` i.e. the number of control points of each Bezier element. This will be problematic if in a multipatch there exist patches of different degree that have the same number of control points (e.g. 2x2 & 4x1). This might be rare but it is not impossible. We could group them by `(PR, PS, PT)` ( the degree per direction) to avoid this.

I have implemented this locally, i.e. I do the grouping with respect to ( NNj, PRj, PSj, PTj). If we agree on this, I can commit it

**Review by filiatra — APPROVED** [review]

**Pull request merged**